examples/mcts.dx

```diff
@@ -0,0 +1,323 @@
+' ## Monte Carlo Tree Search
+MCTS is the decision engine for the notable AlphaGo program.
+It samples rollouts to guide tree search. 
+The nice thing is Dex can in theory automatically parallelize leaf rollouts.
+
+' We implement a variant of MCTS with minimax search and upper
+confidence bound for trees (uct) for the two-player mini-game `Count21`. 
+This demo implements a doubly-linked tree structure as a Dex array,
+and traverses its branches by indexing.
+[reference](http://www.incompleteideas.net/609%20dropbox/other%20readings%20and%20resources/MCTS-survey.pdf)
+
+def (+|) {a} [Ix a] (n:a) (m:Nat) : a = ((ordinal n) + m)@a
+def xor (a:Bool) (b:Bool) : Bool = ((not a) && b) || (a && (not b))
+
+' ### Tree types
+
+data Node node_ix a =
+  Branch content:a parent:(Maybe node_ix) children:(List (node_ix))
+
+def Tree (node_ix:Type) (a:Type) [Ix node_ix] : Type =
+  (node_ix=>(Maybe (Node node_ix a)) & node_ix)
+  -- Node array, and pointer to last used node.
+
+def root_of {node_ix a} (tree:Tree node_ix a) : node_ix = 0@node_ix
+
+def is_leaf {node a} ((Branch _ _ children):Node node a) : Bool = 
+  list_length children == 0
+
+def init_tree {a} (max_size:Nat) (default_content:a): Tree (Fin max_size) a =
+  root = Branch default_content Nothing (AsList 0 [])
+  nothings = for i:(Fin max_size). Nothing
+  init_array = yield_state nothings \ref. ref!(0@_) := (Just root)
+  (init_array, 0@_)
+
+' ### Game and Node utilities
+
+def Game (state:Type) (move:Type) : Type =
+  {
+    possible_moves : state -> (List move) &
+    win_or_lose : state -> Bool &
+    update : state -> move -> state &
+    game_over : state -> Bool &
+  }
+
+def NodeStats (state:Type) : Type = 
+  (Float & Nat & Nat & state)
+  -- (win ratio or reward, #visit, level, state)
```

Comment:
Actually, I guess it's even worse than that: this is the sum across visits over the average across rollouts per visit of (1 for win, 0 for loss).  Though I suppose one could still call it the total reward.

Reply:
Right, I agree that the name was confusing. I updated it so that the first number is actually the number of winning rollouts from that node and its children. Please take another look.